### ISSUE TITLE
Remove account information from returned IdentityCredential

### ISSUE THREAD

**Issue by ekovac** [fedidcg/LightweightFedCM#56]
@samuelgoto has suggested that we should not provide the login-status-stored account information in the IdentityCredential. Effectively, the login-status-stored account information would only be for presentation via the browser-controlled account selection UI, a replacement/precaching mechanism for the `accounts_endpoint` fetch.

This would allow future extension where information on the IdentityCredential could be a verifiable identity assertion, such as in the proposed [Delegation-Oriented FedCM](https://groups.google.com/a/chromium.org/g/blink-dev/c/rwu9wFl0mF4/m/0ADHnMhDCQAJ).

Would like to hear @bvandersloot-mozilla 's thoughts on this.

**Comment by johannhof** [comment]
Sam's suggestion makes sense to me.

**Comment by bvandersloot-mozilla** [comment]
> I'm suggesting that we do not share that metadata to the RP from the `navigator.login.setStatus()` call, but rather, only share that information if the `id_assertion_endpoint` was used.

I don't follow why from your logic. If, as you said, "`navigator.login.setStatus` is a JS client-side alternative for the HTTP server-side accounts_endpoint", and the data is coming from the accounts_endpoint in the FedCM case, why would we not disclose the data from setStatus?

**Comment by samuelgoto** [comment]
> and the data is coming from the accounts_endpoint in the FedCM case

The data that is shared to the RP in the `IdentityCredential` in the full-FedCM case isn't coming the `accounts_endpoint`, it is coming from the `id_assertion_endpoint`.

**Comment by bvandersloot-mozilla** [comment]
> The data that is shared to the RP in the `IdentityCredential` in the full-FedCM case isn't coming from the `accounts_endpoint`, it is coming from the `id_assertion_endpoint`.

Is there a PR # that has this? I can't find it and the Editor's Draft only has the [token and continue_on](https://w3c-fedid.github.io/FedCM/#dictdef-identityassertionresponse) coming from the `id_assertion_endpoint`

**Comment by samuelgoto** [comment]
>  the Editor's Draft only has the token and continue_on coming from the id_assertion_endpoint

The RP gets an [`IdentityCredential`](https://w3c-fedid.github.io/FedCM/#identitycredential): there is nothing that is shared with the RP other than the [`token`](https://w3c-fedid.github.io/FedCM/#dom-identitycredential-token) (which, as you said, comes from `continue_on` or the `id_assertion_endpoint`) and the boolean [`isAutoSelected`](https://w3c-fedid.github.io/FedCM/#dom-identitycredential-isautoselected) (which the browser provides).

The accounts_endpoint is only used to construct an account chooser, but nothing else.

**Comment by bvandersloot-mozilla** [comment]
Ah, I misunderstood- I thought FedCM currently gave the account information on the returned the IdentityCredential via the `CredentialUserData` mixin. If it doesn't +1 from me to remove this as well.

**Comment by samuelgoto** [comment]
> returned the IdentityCredential via the CredentialUserData mixin.

Ah, interesting! It never occurred to me that we could use the mixin! That's great to know actually! But yeah, we don't currently use it at the moment.

> If it doesn't +1 from me to remove this as well.

I just wanted to note that, what I hope to suggest is that, in the **absence** of an `id_assertion_endpoint`, we don't share the `token` in the `IdentityCredential`, but IIUC, in lightweight mode, the `id_assertion_endpoint` is **not** required but **is** optional, so, when one is provided we **should** use it and return a `token`, but when one is **not** we should **not** override that option that the IdP has made and share the data we got from the `navigator.login.setStatus()`.

In other words, setting `navigator.login.setStatus()` should **not** serve as a replacement to `id_assertion_endpoint`.

I think I'm not able to quite express my suggestion precisely, but I hope this is more or less making sense to you all :)

**Comment by johannhof** [comment]
> In other words, setting `navigator.login.setStatus()` should **not** serve as a replacement to `id_assertion_endpoint`.

I agree!

**Comment by ekovac** [comment]
One downside of doing this: It means that if there user has multiple accounts with the IdP, there's no way to determine which one the user selected. Should we populate the `id` field only?

**Comment by npm1** [comment]
You mean adding `account_id` to IdentityCredential or something like that?

**Comment by samuelgoto** [comment]
> One downside of doing this: It means that if the user has multiple accounts with the IdP, there's no way to determine which one the user selected. Should we populate the id field only?

Can you walk me through how you'd expect websites/idps/users to use this flow when there isn't an `id_assertion_endpoint`? Do we actually expect this to happen in practice? Can you give me a concrete example of a developer that might need this and how it would use it?

**Comment by bvandersloot-mozilla** [comment]
The `id` field is inherent to the `Credential` base class, we can't remove that. I think that @ekovac is making sure nobody wants to remove it.

**Comment by cbiesinger** [comment]
That is true but we (Chrome and the spec) do not currently fill it in, i.e. it will be `""`

I guess the proposal is to fill in that field (either for lightweight or both)?

**Comment by bvandersloot-mozilla** [comment]
Interesting! I would have sworn it was set to the `account_id`. That's what I did in Firefox.

**Comment by npm1** [comment]
> Interesting! I would have sworn it was set to the `account_id`. That's what I did in Firefox.

Seems right to me. Any concerns with doing this @samuelgoto ?

**Comment by samuelgoto** [comment]
> Seems right to me. Any concerns with doing this @samuelgoto ?

I don't think I have any concerns passing the `account_id` inside of the `IdenttiyCredential.id` instance (is that's what you are suggestion? or is the suggestion introduce a new `IdentityCredential.accountId` property? -- I guess either could work), but I do have concerns that I don't think I have a clear picture of how this is supposed to work end to end without an `id_assertion_endpoint`:

https://github.com/fedidcg/LightweightFedCM/issues/56#issuecomment-2637856410

That is, who would we expect to call the `navigator.credentials.get()` call? the RP top level frame? or an IdP iframe embedded into an RP top level frame? If the latter, I don't think it matters what we return: the IdP can all SAA after it and `postMessage` up whatever it wants: we could return all of the fields, because they'd be able to get them via `IdentityProvider.getUserInfo()` anyway. 

If the former, how is that supposed to work? Is the RP supposed to `postMessage` down to an IdP iframe to tell it to call SAA? And somewhere in the `postMessage` for the `account_id` to be passed and trusted by the IdP iframe? If so, that sounds a little suspicius to me.

**Comment by cbiesinger** [comment]
For existing FedCM deployments (i.e. pull method) I am a bit wary of exposing new information to JS that it may not have had previously. We can't be sure that the IdP puts the account ID (from the accounts endpoint) into the token (from the ID assertion endpoint). If we want to expose this for the pull method as well, we should talk to some identity providers first.

This is, of course, not an issue for the push method which is all new.

**Comment by ekovac** [comment]
> That is, who would we expect to call the `navigator.credentials.get()` call to get the `IdentityCredential` back allegedly with the newly introduced `account_id`? the RP top level frame? or an IdP iframe embedded into an RP top level frame?
> 
> If the latter, I don't think it matters what we return: the IdP can all SAA after it and `postMessage` up whatever it wants: we could return all of the fields, because they'd be able to get them via `IdentityProvider.getUserInfo()` anyway or third party cookies.
> 
> If the former, how is that supposed to work? Is the RP supposed to `postMessage` down to an IdP iframe to tell it to call SAA? And somewhere in the `postMessage` for the `account_id` to be passed and trusted by the IdP iframe? If so, that sounds a little suspicious to me: how would the IdP iframe trust that the `account_id` was actually what the user selected?

This is what I had in mind but I agree, this is a bit suspicious.

It would be really convenient if the IdP could call some method to at least narrow things down to the set of connected accounts for the user; it looks like getUserInfo() just returns all the accounts from the accounts endpoint. Not being able to support multiple federated logins for the same (RP, IdP) pair if you don't provide an id assertion endpoint seems like possibly an acceptable tradeoff?

**Comment by samuelgoto** [comment]
> The RP top level frame.

Then, what happens? Does the RP top level frame has to `postMessage()` to the IdP iframe so that it can request SAA? and then what?

I'm struggling to find a concrete and real world use case where this would lead to a valid deployment.

**Comment by ekovac** [comment]
Correct; RP top level frame postmessages down to the IdP frame; IdP frame requests SAA, reads the user's account cookies, probably does something that is morally equivalent to making an ID assertion endpoint call, and then postmessages the result of that back up, probably something specific to that IdP.

**Comment by samuelgoto** [comment]
> does something that is morally equivalent to making an ID assertion endpoint call

Wouldn't they be better of exposing an id_assertion_endpoint in this case?

**Comment by ekovac** [comment]
This approach would require less change to an existing third-party-cookie-dependent flow. They might have more frontend eng resources than backend eng resources, or have some specific constraints that means that exposing an ID assertion endpoint that would work correctly with FedCM is just more effort than reusing some existing aspect of their auth flow.

I agree this is mostly hypothetical, but it's certainly not a *useless* configuration.

**Comment by samuelgoto** [comment]
> I agree this is mostly hypothetical, but I think it doesn't cost us anything to support this.

In this setup, we don't have to share an `account_id` in the `IdentityCredential` result, right? If this is a use case that is both hypothetical and that doesn't require us to change anything, is there anything actionable here?

**Comment by ekovac** [comment]
Correct. I'll put together a PR to remove references to reading info off of the IdentityCredential from the explainer.

**Comment by johannhof** [comment]
My POV on this: There are developers today using SAA for identity use cases, where they could, in theory, just as well show a FedCM prompt (maybe active mode). For these developers, I want to make it as easy as possible to use the right UI paradigm (FedCM) instead.

**Comment by ekovac** [comment]
Fixed with #60
